### ISSUE TITLE
In v0.32.0, `.json()` returns an empty string if "Transfer-Encoding: chunked" header is present (and thus "Content-Length: [...]" header is absent)

### ISSUE THREAD

**Issue by davidrunger** [sindresorhus/ky#463]
This bug was introduced in https://github.com/sindresorhus/ky/pull/459 (cc @lemonadern, for your interest).

**Comment by davidrunger** [comment]
^ there is linked a PR (https://github.com/sindresorhus/ky/pull/464) that should fix this bug

**Comment by Nickhoyer** [comment]
@davidrunger @sindresorhus Compressed Cloudflare responses come without Content-Length and without Transfer-Encoding

https://community.cloudflare.com/t/no-content-length-header-when-response-is-compressed/134475

**Comment by davidrunger** [comment]
@Nickhoyer Good catch. That is a problem.

My PR #464 fixed the bug that was failing my app's test suite... and so now that v0.32.1 has passed my test suite, it's in my app out in production behind Cloudflare, and I can confirm that I am indeed seeing this error (`Uncaught (in promise) TypeError: t.forEach is not a function`) on the client-side in production now, because Cloudflare is indeed providing neither a `Content-Length` nor a `Transfer-Encoding` header.

It would probably be good to come up with a solution ASAP, before too many people ship this "occurs only in production (behind Cloudflare)" bug to users. 😬

**Comment by davidrunger** [comment]
> It seems to me that a request can provide a content-encoding header for a stream instead of a transfer-encoding header for chunks

So basically the idea would be to add another conditional, then?, i.e.:

```js
const contentLength = response.headers.get('Content-Length');
const contentEncoding = response.headers.get('Content-Encoding');
const transferEncoding = response.headers.get('Transfer-Encoding');

const responseIsEmpty =
  (contentLength === null || contentLength === '0') &&
    contentEncoding === null &&
    transferEncoding !== 'chunked'

if (responseIsEmpty) {
  return '';
}
```

Are we confident that would have the desired effect?, i.e. will `Content-Encoding` only be present if the response is not empty?

If we're not confident of that, then the best other solution that I can think of in the meantime would be to get rid of this code entirely for now (i.e. revert both #464 and #459).

**Comment by davidrunger** [comment]
^ linked there is a PR (https://github.com/sindresorhus/ky/pull/465) implementing the code change that I proposed in that comment above. I haven't been able to add a test to that PR yet (and might never), nor have I been able to test that fix in production behind Cloudflare yet, but I wanted to put that up ASAP in case @sindresorhus wants to merge something in quickly to prevent this bug from popping up only after folks have shipped it out to production (where Cloudflare adjusts the headers that we are currently relying on).

**Comment by davidrunger** [comment]
@0f-0b That is good to know. In addition to the service worker case that you mentioned, although my new fix attempt #465 did fix one occurrence of the error in question in production, I recently managed to find a different request in my web app in production behind Cloudflare that for whatever reason* does not have any of the three headers which were mentioned above (`'Content-Length'`, `'Content-Encoding'`, or `'Transfer-Encoding'`), and so that request is still erroring, even though I have deployed the #465 branch to production.

In summary, it seems that, as you suggested, trying to use the response headers to determine whether there is body content is probably never going to be sufficiently reliable.

---

*I think maybe because the JSON response is too short for Cloudflare to bother compressing it, so there's no `Content-Encoding' header, and yet Cloudflare also apparently still strips off the `Content-Length` header

**Comment by patrick-entinux** [comment]
Here is an example of a public URL that erroneously returns empty string in v0.32 too:

```
ky('https://cdn.jsdelivr.net/npm/jquery@3.2.1/package.json').json();
```

Response headers as reported by curl:

<details>
<summary>curl -I -X GET https://cdn.jsdelivr.net/npm/jquery@3.2.1/package.json</summary>

```
date: Tue, 01 Nov 2022 09:20:33 GMT
content-type: application/json; charset=utf-8
access-control-allow-origin: *
access-control-expose-headers: *
timing-allow-origin: *
cache-control: public, max-age=31536000, s-maxage=31536000, immutable
cross-origin-resource-policy: cross-origin
x-content-type-options: nosniff
strict-transport-security: max-age=31536000; includeSubDomains; preload
x-jsd-version: 3.2.1
x-jsd-version-type: version
etag: W/"931-BtsGBIZnvkQ5OzTQ5/2nU1JyrIg"
x-served-by: cache-fra19145-FRA, cache-yyz4536-YYZ
x-cache: HIT, HIT
vary: Accept-Encoding
alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
cf-cache-status: HIT
age: 1387
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=x46gIJvHFIKvwDQ%2Fqmu%2B5j73ri732M%2FOrUo2IKoSNSlzsUHqG1OVLi6PPFsH9%2Fq9EYiKIEPcili5RiOrsCfIRmHmS%2BWDVteMGkHhuiQgyEMxo4lIozy8CbtRsovkYeLc%2FT%2BKAlPJWCP%2BNPZyJrs%3D"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0.01,"report_to":"cf-nel","max_age":604800}
server: cloudflare
cf-ray: 76338e220850838b-KIX
```

</details>

<details>
<summary>curl --compressed -I -X GET https://cdn.jsdelivr.net/npm/jquery@3.2.1/package.json</summary>

```
date: Tue, 01 Nov 2022 09:20:23 GMT
content-type: application/json; charset=utf-8
access-control-allow-origin: *
access-control-expose-headers: *
timing-allow-origin: *
cache-control: public, max-age=31536000, s-maxage=31536000, immutable
cross-origin-resource-policy: cross-origin
x-content-type-options: nosniff
strict-transport-security: max-age=31536000; includeSubDomains; preload
x-jsd-version: 3.2.1
x-jsd-version-type: version
etag: W/"931-BtsGBIZnvkQ5OzTQ5/2nU1JyrIg"
x-served-by: cache-fra19145-FRA, cache-yyz4536-YYZ
x-cache: HIT, HIT
vary: Accept-Encoding
alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
cf-cache-status: HIT
age: 1377
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=rDocGZrL5nt%2FzljxqXez5zUNBab8tiwLFpzv%2BCDRF%2Bcc%2B8Gunv4eYYKZsC%2FUcpAqSg0j8VWGwmHo9pFOb5MatOKdktDvOb3kAh2Vu4hAuPXHJ2nDiWE5mRqJWa7wFwp%2B%2F3p86wA7YmW1gThFzKY%3D"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0.01,"report_to":"cf-nel","max_age":604800}
server: cloudflare
cf-ray: 76338de01c4a836a-KIX
content-encoding: br
```
</details>

**Comment by sindresorhus** [comment]
@patrick-entinux Can you try out 0.32.2?

**Comment by patrick-entinux** [comment]
@sindresorhus I tried 0.32.2 and it works great! Fixes my issues completely.
